msprime/ancestry.py

```diff
@@ -793,59 +756,69 @@ def _insert_samples_from_map(samples_map, demography, ploidy, tables):
+            raise TypeError(
+                "The number of samples to draw from a population must be an integer"
             )
+        sample_set.num_samples = int(sample_set.num_samples)
+        if sample_set.num_samples < 0:
+            raise ValueError("Number of samples cannot be negative")
+        if sample_set.population is None:
+            if demography.num_populations == 1:
+                sample_set.population = 0
+            else:
+                raise ValueError(
+                    "Must specify a SampleSet population in multipopulation models"
+                )
+        else:
+            sample_set.population = _get_population(demography, sample_set.population)
+
+    if sum(sample_set.num_samples for sample_set in sample_sets) == 0:
+        raise ValueError("Zero samples specified")
+    return sample_sets
 
 
 def _parse_samples(samples, demography, ploidy, tables):
     """
-    Parse the specified "samples" value for sim_ancesty and insert them into the
+    Parse the specified "samples" value for sim_ancestry and insert them into the
     specified tables.
     """
-    samples_list = None
-    samples_map = None
     if isinstance(samples, collections.abc.Sequence):
-        samples_list = _parse_samples_list(samples, demography)
+        sample_sets = samples
     elif isinstance(samples, collections.abc.Mapping):
-        samples_map = _parse_samples_map(samples, demography)
-    elif isinstance(samples, numbers.Number):
-        if not core.isinteger(samples):
-            raise ValueError("The number of samples must be an integer")
-        n = int(samples)
+        sample_sets = [
+            SampleSet(num_samples, population)
+            for population, num_samples in samples.items()
+        ]
+    elif core.isinteger(samples):
         if len(tables.populations) != 1:
             raise ValueError(
                 "Numeric samples can only be used in single population models. "
                 "Please use Demography.sample() to generate a list of samples "
                 "for your model, which can be used instead."
             )
-        samples_map = {0: n}
+        sample_sets = [SampleSet(samples)]
     else:
-        # TODO better error message, like this one
-        # raise TypeError(
-        #     "The samples argument must be either an integer (for single population "
-        #     "models) or a mapping from populations to sample numbers/time values"
-        # )
-        raise TypeError("bad samples type")
-    if samples_map is not None:
-        assert samples_list is None
-        _insert_samples_from_map(samples_map, demography, ploidy, tables)
-    else:
-        assert samples_list is not None
-        _insert_samples_from_list(samples_list, demography, ploidy, tables)
+        raise TypeError(
+            f"The value '{samples}' cannot be interpreted as sample specification. "
+            "Samples must either be a single integer, a mapping of populations "
```

Comment:
maybe "a dict that maps" rather than "a mapping"

Reply:
Well it doesn't have to be a dict (could be, say, a collections.defaultdict), but you're right it's easier to follow if it's concrete.